design/01-proc-macro.md

```diff
@@ -0,0 +1,290 @@
+In contrast to current behaviour of Cairo plugins, no longer will they be executed on each node of AST.
+
+The procedural macro will be triggered by one of three Cairo expressions
+- Macro call, e.g. `xyz!`
+- Macro attribute, e.g. `#[xyz]`
+- Macro derive, e.g. `#[derive(Xyz)]`
+
+**Example:**
+
+Scarb manifest file:
+
+```toml
+[package]
+name = "hello-macros"
+version = "0.1.0"
+
+[dependencies]
+add-macro = "0.1.0"
+tracing-macro = "0.1.0"
+to-value-macro = "0.1.0"
+```
+
+Cairo source file:
+
+```cairo
+use add_macro::add;
+use tracing_macro::instrument;
+use to_value_macro::ToValue;
+
+#[derive(ToValue)]
+struct Input {
+    value: felt252, 
+}
+
+#[instrument]
+fn main() -> felt252 {
+    let a = Input { value: 1 };
+    let b = Input { value: 2 };
+    add!(a.to_value(), b.to_value());
+}
+```
+
+### Procedural macro author perspective
+
+To implement a procedural macro, a programmer will have to:
+- Create a new package, with a `Scarb.toml` manifest file and a `src/` directory besides.
+- The manifest file must define a `cairo-plugin` target type.
+- Write a Rust library, inside the `src/` directory, that implements the procedural macro API.
+- Note, that the package does not contain any `Cargo.toml` file, as it will be generated by Scarb.
+- A Rust crate exposing an API for writing procedural macros will be published for programmers. This crate will be automatically added to generated `Cargo.toml` file.
+- The Rust library contained in the package will have to implement a public `generate_code` function.
+- This function accepts a `TokenStream`  as an input and returns a `ProcMacroResult` as an output, both defined in the helper library.
+- The result enum will contain one of three values:
+    - `ProcMacroResult::Leave` - procedural macro does not take any action
+    - `ProcMacroResult::Replace({ TokenStream, AuxData })` - procedural macro has generated code that replaces input
+    - `ProcMacroResult::Remove` - original code should be removed
+    - `ProcMacroResult::Bail(ProcMacroDiagnostics)` - procedural macro exited with an error
+- Alongside the new TokenStream, a procedural macro can emit auxiliary data, encoded as an arbitrary JSON.
+
+A `TokenStream` definition:
+- A token can be defined as a string with one of three kinds assigned to it during parsing.
+- The token kinds are: identifier, punctuation symbol and literal.
+- Kinds of different tokens can be represented as an enum called `TokenTree`.
+- An iterator over token trees is called a `TokenStream`. A `TokenStream` is a representation of a Cairo code.
+- The `TokenStream` are plain Cairo tokens that will be parsed by the procedural macro implementation itself.
+
+Example:
+
+```toml
+[package]
+name = "some-macro"
+version = "0.1.0"
+
+[cairo-plugin]
+```
+
+```rust
+/// The entry point of procedural macro implementation.
+pub fn generate_code(token_stream: TokenStream) -> ProcMacroResult {
+    // no-op
+    ProcMacroResult::Leave
+}
```

Comment:
```suggestion
use cairo-lang-macro::{inline_macro, ProcMacroResult, TokenStream};

/// The entry point of procedural macro implementation.
#[inline_macro]
pub fn some_macro(token_stream: TokenStream) -> ProcMacroResult {
    // no-op
    ProcMacroResult::Leave
}
```

1. Let's emphasize usage of the API crate.
2. This outline does not say how macro name and kind is determined.
3. Without `extern "C"` Rust will mangle the `generate_code` name, so it wasn't a good FFI entrypoint. Wrapping this into custom procmacro will allow us to hide this implementation detail.

Reply:
Actually, wouldn't it be easier to use package name as macro name? 🤔

Reply:
no, because I imagine it would be popular to have single package exposing a collection of related macros

Reply:
I understand this would be nice to have, but it would also open other problems - like we could no longer rely on scarb resolve to ensure no two macros with same name are declared.

Reply:
I think an ability to easily bundle multiple macros together, with shared state and code etc. is much bigger benefit than simplifying duplicate names check (which actually should boil down to just looping over all loaded plugins before feeding them into the DB as far as I imagine)